File: app/android/app/src/main/java/app/walkpedia/WalkPediaDatabase.kt
package app.walkpedia

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import java.io.File

/**
 * Read-only access to the bundled `walkpedia.sqlite`.
 *
 * On first launch the asset is copied into app storage (SQLiteDatabase needs a
 * real file path). The nearest-N query uses the R-tree index when available, and
 * transparently falls back to a plain-index bounding-box query on devices whose
 * system SQLite was built without the R-tree module.
 */
class WalkPediaDatabase private constructor(private val db: SQLiteDatabase) {

    private val hasRtree: Boolean by lazy {
        try {
            db.rawQuery("SELECT page_id FROM article_rtree LIMIT 1", null).use { it.moveToFirst() }
            true
        } catch (e: Exception) {
            false
        }
    }

    fun nearest(lat: Double, lon: Double, radiusMeters: Double, limit: Int): List<Article> {
        val dLat = radiusMeters / 111_320.0
        val dLon = radiusMeters / (111_320.0 * maxOf(Math.cos(Math.toRadians(lat)), 1e-6))
        val minLat = lat - dLat; val maxLat = lat + dLat
        val minLon = lon - dLon; val maxLon = lon + dLon

        val (sql, args) = if (hasRtree) {
            """
            SELECT a.page_id,a.title,a.lat,a.lon,a.type,a.summary
            FROM article_rtree r JOIN article a ON a.page_id = r.page_id
            WHERE r.max_lat>=? AND r.min_lat<=? AND r.max_lon>=? AND r.min_lon<=?
            """.trimIndent() to arrayOf(maxLat, minLat, maxLon, minLon)
        } else {
            """
            SELECT page_id,title,lat,lon,type,summary FROM article
            WHERE lat BETWEEN ? AND ? AND lon BETWEEN ? AND ?
            """.trimIndent() to arrayOf(minLat, maxLat, minLon, maxLon)
        }

        val scored = ArrayList<Pair<Double, Article>>()
        db.rawQuery(sql, args.map { it.toString() }.toTypedArray()).use { c ->
            while (c.moveToNext()) {
                val alat = c.getDouble(2)
                val alon = c.getDouble(3)
                val d = haversine(lat, lon, alat, alon)
                if (d <= radiusMeters) {
                    scored.add(
                        d to Article(
                            id = c.getLong(0),
                            title = c.getString(1) ?: "(untitled)",
                            lat = alat, lon = alon,
                            type = c.getString(4),
                            summary = c.getString(5) ?: ""
                        )
                    )
                }
            }
        }
        return scored.sortedBy { it.first }.take(limit).map { it.second }
    }

    companion object {
        private const val ASSET = "walkpedia.sqlite"

        fun open(context: Context): WalkPediaDatabase {
            val out = File(context.filesDir, ASSET)
            if (!out.exists()) {
                context.assets.open(ASSET).use { input ->
                    out.outputStream().use { input.copyTo(it) }
                }
            }
            val db = SQLiteDatabase.openDatabase(out.path, null, SQLiteDatabase.OPEN_READONLY)
            return WalkPediaDatabase(db)
        }
    }
}

File: app/ios/WalkPedia/WalkPediaDatabase.swift
import Foundation
import SQLite3

/// One geotagged Wikipedia article from the bundled WalkPedia dataset.
struct Article: Identifiable, Equatable {
    let id: Int64            // Wikipedia page_id
    let title: String
    let lat: Double
    let lon: Double
    let type: String?        // landmark | city | mountain | ...
    let summary: String      // TTS-ready plain-text lead
}

/// Read-only access to the bundled `walkpedia.sqlite` (built by the data pipeline).
/// Implements the core "N closest within R metres" query using the R-tree index
/// for a cheap bounding-box pre-filter, then haversine for metre-accurate ranking.
final class WalkPediaDatabase {
    private var db: OpaquePointer?

    init?(bundledName: String = "walkpedia", ext: String = "sqlite") {
        guard let url = Bundle.main.url(forResource: bundledName, withExtension: ext) else {
            print("WalkPedia: \(bundledName).\(ext) missing from app bundle")
            return nil
        }
        if sqlite3_open_v2(url.path, &db, SQLITE_OPEN_READONLY, nil) != SQLITE_OK {
            print("WalkPedia: could not open database")
            return nil
        }
    }

    deinit { sqlite3_close(db) }

    /// The `limit` closest articles within `radiusMeters` of (lat, lon), nearest first.
    func nearest(lat: Double, lon: Double, radiusMeters: Double, limit: Int) -> [Article] {
        let dLat = radiusMeters / 111_320.0
        let dLon = radiusMeters / (111_320.0 * max(cos(lat * .pi / 180), 1e-6))
        let sql = """
            SELECT a.page_id, a.title, a.lat, a.lon, a.type, a.summary
            FROM article_rtree r JOIN article a ON a.page_id = r.page_id
            WHERE r.max_lat >= ? AND r.min_lat <= ?
              AND r.max_lon >= ? AND r.min_lon <= ?
            """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return [] }
        defer { sqlite3_finalize(stmt) }
        sqlite3_bind_double(stmt, 1, lat - dLat)
        sqlite3_bind_double(stmt, 2, lat + dLat)
        sqlite3_bind_double(stmt, 3, lon - dLon)
        sqlite3_bind_double(stmt, 4, lon + dLon)

        var scored: [(Double, Article)] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            let title = sqlite3_column_text(stmt, 1).map { String(cString: $0) } ?? "(untitled)"
            let alat = sqlite3_column_double(stmt, 2)
            let alon = sqlite3_column_double(stmt, 3)
            let d = WalkPediaDatabase.haversine(lat, lon, alat, alon)
            guard d <= radiusMeters else { continue }
            let article = Article(
                id: sqlite3_column_int64(stmt, 0),
                title: title,
                lat: alat, lon: alon,
                type: sqlite3_column_text(stmt, 4).map { String(cString: $0) },
                summary: sqlite3_column_text(stmt, 5).map { String(cString: $0) } ?? ""
            )
            scored.append((d, article))
        }
        return scored.sorted { $0.0 < $1.0 }.prefix(limit).map { $0.1 }
    }

    /// Great-circle distance in metres.
    static func haversine(_ lat1: Double, _ lon1: Double, _ lat2: Double, _ lon2: Double) -> Double {
        let R = 6_371_000.0
        let p1 = lat1 * .pi / 180, p2 = lat2 * .pi / 180
        let dp = (lat2 - lat1) * .pi / 180, dl = (lon2 - lon1) * .pi / 180
        let a = sin(dp / 2) * sin(dp / 2) + cos(p1) * cos(p2) * sin(dl / 2) * sin(dl / 2)
        return 2 * R * asin(min(1, sqrt(a)))
    }
}
